### PR TITLE
 Document how users can use pytest instead of kedro test

### PR DESCRIPTION
**Pull request by jmholzer**
## Description
Resolves [#1619](https://github.com/kedro-org/kedro/issues/1619). This PR provides basic documentation on how to install, configure and use `pytest` and `pytest-cov` to provide automating testing for a Kedro project.

How to write tests specifically for Kedro is not in the scope of this PR, this is the scope of the issue https://github.com/kedro-org/kedro/issues/1271, which is related but not the same. I see the opportunity to add the documentation for this open issue in the same place as the new documentation added by this PR.

## Development notes
I added the file `automated_testing.md`, which contains the newly written documentation and added an entry for it under `index.rst` under the 'development' section.

I have enclosed the names of third-party libraries in backticks (e.g. `pytest-cov`, `coverage.py`). In two cases these names are also links to the documentation for these libraries. This is a stylistic choice in the docs of other projects, but I could not immediately see if we do it here in Kedro, this could be wrong.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1879"></a>

### CODE REVIEW

File: docs/source/development/automated_testing.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    def test_project_path(self, project_context):
+        assert project_context.project_path == Path.cwd()
+```
+
+This test is redundant, but it introduces a few of `pytest`'s core features and demonstrates the layout of a test file:
+- [Fixtures](https://docs.pytest.org/en/7.1.x/explanation/fixtures.html#about-fixtures) are used to define resources used in tests.
+- Tests are implemented in methods or functions beginning with `test_` and classes beginning with `Test`.
+- The `assert` statement is used to compare the result of the test with an expected value.
+
+You can read more about the basics of using `pytest` [here](https://docs.pytest.org/en/7.1.x/getting-started.html). For help writing your own tests and using all of the features of `pytest`, see the [docs](https://docs.pytest.org/).
```
Comment (MerelTheisenQB) [inline docs/source/development/automated_testing.md:95]:
```suggestion
You can read more about the [basics of using `pytest` on the getting started page] (https://docs.pytest.org/en/7.1.x/getting-started.html). For help writing your own tests and using all of the features of `pytest`, see the [pytest docs](https://docs.pytest.org/).
```

Making the links a bit more descriptive for accessibility: https://github.com/kedro-org/kedro/issues/1281

File: docs/source/development/automated_testing.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+This output indicates that one test ran successfully in the file `src/tests/test_run.py`.
+
+## Add test coverage reports with `pytest-cov`
+
+It can be useful to see how much of your project is covered by tests. For this, you can install and configure the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin for `pytest`, which is based on the popular [`coverage.py`](https://coverage.readthedocs.io/) library.
+
+### Install `pytest-cov`
+
+Install `pytest` as you would install other packages with pip, making sure your project's virtual environment is active.
```
Comment (MerelTheisenQB) [inline docs/source/development/automated_testing.md:126]: Could link to the virtual env setup docs again here.

File: docs/source/development/automated_testing.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+src/spaceflights/pipelines/data_processing/pipeline.py       5      5     0%   1-8
+src/spaceflights/pipelines/data_science/__init__.py          1      1     0%   3
+src/spaceflights/pipelines/data_science/nodes.py            20     20     0%   1-55
+src/spaceflights/pipelines/data_science/pipeline.py          8      8     0%   1-40
+src/spaceflights/settings.py                                 0      0   100%
+--------------------------------------------------------------------------------------
+TOTAL                                                       98     98     0%
+```
+
+This is the simplest report that `coverage.py` (via `pytest-cov`) will produce. It gives an overview of how many of the executable statements in each project file are covered by tests. For a full set of features offered by `coverage.py`, see the [docs](https://coverage.readthedocs.io/).
```
Comment (MerelTheisenQB) [inline docs/source/development/automated_testing.md:165]:
```suggestion
This is the simplest report that `coverage.py` (via `pytest-cov`) will produce. It gives an overview of how many of the executable statements in each project file are covered by tests. For a full set of features offered by `coverage.py`, see the [coverage docs](https://coverage.readthedocs.io/).
```

**Review by MerelTheisenQB — APPROVED** [review]
These docs are absolutely brilliant!! 🤩  So well written! I only have some minor comments. 
One slightly bigger comment is on linking this page from other parts of the documentation so users can find it. Maybe we can add a section in "get started" around good software engineering practices and then link this page as well as the linting instructions.

**Review by AhdraMeraliQB — APPROVED** [review]
These are amazingly well-written, great job! 🌟

File: docs/source/development/automated_testing.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+This test is redundant, but it introduces a few of `pytest`'s core features and demonstrates the layout of a test file:
+- [Fixtures](https://docs.pytest.org/en/7.1.x/explanation/fixtures.html#about-fixtures) are used to define resources used in tests.
+- Tests are implemented in methods or functions beginning with `test_` and classes beginning with `Test`.
+- The `assert` statement is used to compare the result of the test with an expected value.
+
+You can read more about the basics of using `pytest` [here](https://docs.pytest.org/en/7.1.x/getting-started.html). For help writing your own tests and using all of the features of `pytest`, see the [docs](https://docs.pytest.org/).
+
+### Run your tests
+
+To run your tests, run `pytest` from within your project's root directory or from the `/src` directory.
```
Comment (noklam) [inline docs/source/development/automated_testing.md:99]: Maybe we don't need need to mention the `src/`. In theory, `pytest` would still work, but the following section we also mention `pyproject.toml`, and it won't be picked up if it's running from `src/`

File: docs/source/development/automated_testing.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+Now that `pytest` is installed, you will need a place to put your tests. Create a `/tests` folder in the `/src` directory of your project.
+
+```bash
+mkdir <project_root>/src/tests
+```
+
+### Test directory structure
+
+The subdirectories in your project's `/tests` directory should mirror the directory structure of your project's `/src/<project_name>` directory. All files in the `/tests` folder should be named `test_<file_being_tested>.py`. For instance:
```
Comment (noklam) [inline docs/source/development/automated_testing.md:40]:
```suggestion
The subdirectories in your project's `/tests` directory should mirror the directory structure of your project's `/src/<package_name>` directory. All files in the `/tests` folder should be named `test_<file_being_tested>.py`. For instance:
```

We should stick with this naming convention.  https://github.com/kedro-org/kedro/pull/1753

**Review by noklam — APPROVED** [review]
Nicely written! I drop some comments, but it looks very good already. 

We should stick with <package_name> and <project-name> terminology that we established here https://github.com/kedro-org/kedro/pull/1753.

File: docs/source/development/automated_testing.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+## Set up automated testing with `pytest`
+
+There are many testing frameworks available for Python. One of the most popular is [`pytest`](https://docs.pytest.org/). `pytest` is often used in Python projects for its short, readable tests and powerful set of features.
+
+Let's look at how you can start working with `pytest` in your Kedro project.
+
+### Install `pytest`
+
+Install `pytest` as you would install other packages with `pip`, making sure your project's virtual environment is active.
```
[the remark this answers is not in the document]
Comment (jmholzer) [reply docs/source/development/automated_testing.md]: Great idea, I've added this in [965c6fe](https://github.com/kedro-org/kedro/pull/1879/commits/965c6fe777c0a640e4ee3014dd5c75df800966d9).

File: docs/source/development/automated_testing.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    )
+
+class TestProjectContext:
+    def test_project_path(self, project_context):
+        assert project_context.project_path == Path.cwd()
+```
+
+This test is redundant, but it introduces a few of `pytest`'s core features and demonstrates the layout of a test file:
+- [Fixtures](https://docs.pytest.org/en/7.1.x/explanation/fixtures.html#about-fixtures) are used to define resources used in tests.
+- Tests are implemented in methods or functions beginning with `test_` and classes beginning with `Test`.
```
[the remark this answers is not in the document]
Comment (jmholzer) [reply docs/source/development/automated_testing.md:92]: Yep! Added in [ceecd5f](https://github.com/kedro-org/kedro/pull/1879/commits/ceecd5f2efd16997a2553a961174e9faccd57e7e)

File: docs/source/development/automated_testing.md (hunk 1 above)
Comment (jmholzer) [reply docs/source/development/automated_testing.md]: You're completely right about this. I fixed this in multiple places in [0a5795c](https://github.com/kedro-org/kedro/pull/1879/commits/0a5795c4e800d6f8a63a538470564d562707126d).

File: docs/source/development/automated_testing.md (hunk 2 above)
Comment (jmholzer) [reply docs/source/development/automated_testing.md]: Added in [13586b6](https://github.com/kedro-org/kedro/pull/1879/commits/13586b61d8d9c81d5196e0358ee010ccb4e710e7).

File: docs/source/development/automated_testing.md (hunk 3 above)
Comment (jmholzer) [reply docs/source/development/automated_testing.md]: Fixed in [8650a0b](https://github.com/kedro-org/kedro/pull/1879/commits/8650a0bd5726382b78f1eeaa7439d121ac297ae3).

File: docs/source/development/automated_testing.md (hunk 5 above)
Comment (jmholzer) [reply docs/source/development/automated_testing.md]: You're right about this. I changed three instances of <project_name> to <package_name> in [6a6a278](https://github.com/kedro-org/kedro/pull/1879/commits/6a6a2788d43e93c554e0e2381636df0ff1d366d7).

**Comment by jmholzer** [comment]
Thanks so much for the very insightful and helpful reviews 🎉. 

@MerelTheisenQB I like your idea concerning a new section under 'Get Started', should we open a new ticket for this and create a PR for it after the changes for the new linting docs ([#1620](https://github.com/kedro-org/kedro/issues/1620)) have also been merged?

**Comment by jmholzer** [comment]
> Yes that's a good idea! 👍

Great! I'll get on that 🙂. Thanks for the review.

**Pull request merged**